Firestore/core/src/firebase/firestore/model/document_key.mm

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+#if defined(__OBJC__)
```

Comment:
This doesn't work the way you want: `.mm` files are always compiled as Objective-C++. This `#if` will always be true.

The better way to set this up is with constructors and user-defined conversion operators in C++.
This:
  * Allows you to avoid creating `document_key.mm`, keeping the C++ side pure
  * Avoids the `#include` loop you were seeing
  * Has the compiler perform the necessary conversions for you

Concretely the way to do this is:

  * In `document_key.h`:
    ```c++
    #if __OBJC__
    #import "Firestore/Source/Model/FSTDocumentKey.h"
    #endif  // __OBJC__
    ```
    And in the definition of `DocumentKey`, add a constructor that performs the conversion. If you don't mark it explicit, the compiler will automatically perform the conversion for you:
    ```c++
    #if __OBJC__
    DocumentKey(FSTDocumentKey* key) : path_(key.path) {
    }
    #endif
    ```
    You can also define a conversion operator too:
    ```
    #if __OBJC__
    operator FSTDocumentKey*() const {
      return [FSTDocumentKey keyWithPath:path()];
    }
    #endif
    ```
  * `document_key.cc` no longer has to know anything about Objective-C++ and can just be a regular C++ file.
  * `FSTDocumentKey.h` and `FSTDocumentKey.mm` no longer have to do anything special so there's no `#include` loop.

Reply:
done